docs/tutorial/creating_cli.md

```diff
@@ -0,0 +1,622 @@
+
+```scala 3
+override def searchId(username: Username): IO[Option[Contact]] =
+  for 
+    contacts <- bookPath.readLines
+  yield contacts
+    .find(_.startsWith(username))
+    .flatMap(parseContact(_).toOption)
+```
+
+Similar to read, we load all the contacts from the file, and then we find the contact that matches the username we want to search using the `find` functions, it returns an `Option` whenever it found something or not. If we find it, we parse it to a `Contact` and bind it as an `Option`.
+
+The next three functions are very similar, so we will only show one of them, the `searchName` function:
+
+```scala 3
+override def searchName(name: Name): IO[List[Contact]] =
+  for
+    contacts <- bookPath.readLines.map(_.filter(_.contains(name)))
+    parsed <- contacts.traverse(contact => 
+      IO.fromEither(parseContact(contact)) 
+    )
+  yield parsed
+```
+
+The first part is similar to the previous functions, we load all the contacts from the file, and then we filter the ones that contain the name (or field) we want to search. Look! It's our friend `traverse` again! We're going to use it to parse the file representation of the contacts to our `Contact`using the `fromEither` method, as this will raise the error if it encounters a `Left` inside the `IO`.
+
+Next, we have the `getAll` function to list all the contacts:
+
+```scala 3
+override def getAll: IO[List[Contact]] =
+  for
+    contacts <- bookPath.readLines
+    parsed <- contacts.traverse(contact =>
+      IO.fromEither(parseContact(contact))
+    )
+  yield parsed
+```
+In like the search functions, we load all the contacts from the file, and then we parse them to a `Contact` using the `traverse` method.
+
+The last function we need to implement is the `updateContact` function, be aware that this is the most complex function to implement, so we will break it down into smaller parts.
+
+```scala 3
+override def updateContact(
+    username: Username
+)(modify: Contact => Contact): IO[Contact] =
+  bookPath
+    .readLines
+    .flatMap(_.toVector.traverse(c => IO.fromEither(parseContact(c)))) // (1)
+    .flatMap: contacts =>
+      contacts.zipWithIndex.find(_._1.username == username) match      // (2)
+        case Some((contact, index)) =>
+          val updated = modify(contact)
+          bookPath.writeLines(                                         // (3)              
+            contacts.updated(index, updated).map(encodeContact)        // (4)
+          ) *> IO.pure(updated)
+        case None =>
+          new Exception(s"Contact $username nor found")
+            .raiseError[IO, Contact]
+      
+    
+```
+
+1. We convert all the contacts from the file by parsing them to a `Contact` using the `traverse` method along with the `fromEither` method. Note that we also transform the `List` to a `Vector`, this is because a Vector has better performance when we need to read or update elements in random positions. 
+
```

Comment:
Here also it reads as if `traverse` and `fromEither` do the parsing, but they actually don't, they make it more ergonomic to work with types.